typing/typedecl_unboxed.ml

```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+              Path.print p
+          else match Env.find_type_descrs p env, Env.find_type p env with
+          | exception Not_found -> any
+          | descr, decl ->
+              of_typedescr env descr decl ~args callstack_map
+                (Callstack.visit p head_callstack)
+        end
+    | Ttuple _ -> block_shape [0]
+    | Tarrow _ | Tpackage _ | Tobject _ | Tnil | Tvariant _ -> (* XXX *) any
+    | Tlink _ | Tsubst _ | Tpoly _ | Tfield _ ->
+        assert false
+
+
+  and of_typedescr env ty_descr ty_decl ~args
+                   callstack_map new_callstack =
+    (* we called scrape_ty on the type expression in the function above, thus
+       it isnt a single constructor unboxed type.
+       TODO check what happens in the [@@unboxed] case (on the whole definition)
+     *)
+    let of_type_with_params ty =
+      (* We instantiate the formal type variables with the
+         type expression parameters at use site.
+
+         We build a callstack for this new type by mapping all new
+         nodes, corresponding to the spine of 'ty', to the current
+         call stack. *)
+      let ty = ty_of_poly ty in
+      let params = ty_decl.type_params in
+      let ty = Ctype.apply env params ty args in
+      let callstack_map =
+        Callstack.extend callstack_map ty new_callstack in
+      of_type env ty callstack_map
+    in
+    match ty_descr with
+    | Type_abstract ->
+       begin match ty_decl.type_manifest with
+       | None -> any
+       | Some ty -> of_type_with_params ty
+       end
+    | Type_record (_, Record_regular) -> block_shape [0]
+    | Type_record (_, Record_float) -> block_shape [Obj.double_array_tag]
+    | Type_record ([{lbl_arg = ty; _}], Record_unboxed _) ->
+        of_type_with_params ty
+    | Type_record (_, Record_unboxed _)
+    | Type_record (_, Record_inlined _)
+    | Type_record (_, Record_extension _) -> assert false
+    | Type_open -> block_shape [0]
+    | Type_variant ([],_) -> none
+    | Type_variant ((fst_descr :: _) as cstr_descrs,_) ->
+        (* the head shape of boxed constructors is equivalent to the nb of
+           constant constructors and the nb of non constant constructors *)
+        let num_consts = fst_descr.cstr_consts in
+        let num_nonconsts = fst_descr.cstr_nonconsts in
+        let boxed_shape = cstrs_shape ~num_consts ~num_nonconsts in
+        let unboxed_shapes = List.filter_map
+          (fun descr ->
+            match descr.cstr_tag with
+            | Cstr_constant _ | Cstr_block _ | Cstr_extension _ -> None
+            | Cstr_unboxed {cache_ty; _}->
+                Some (of_type_with_params cache_ty)
+          ) cstr_descrs
+        in
+        (* now checking that the unboxed constructors are compatible with the
+           base head shape of boxed constructors *)
+        List.fold_left disjoint_union boxed_shape unboxed_shapes
+
+  let fill_cache env {cache_ty; cache_shape} =
```

Comment:
The name is correct, but `get`` would also work and may read better at the callsite.

(Or: have a `fill_cache` version that returns `unit`, and a `get` version that fills the cache and returns the head shape. `fill_cache head_shape_data` reads better than `ignore (get head_shape_data)`.)